core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardsStoresAction.java

```diff
@@ -0,0 +1,255 @@
+                ImmutableList.Builder<IndicesShardsStoresResponse.Failure> failures = ImmutableList.builder();
+                while (!fetchResponses.isEmpty()) {
+                    Response res = fetchResponses.remove();
+                    IntObjectHashMap<java.util.List<NodeGatewayStartedShards>> indexShards = shardsResponses.get(res.shardId.getIndex());
+                    final IntObjectHashMap<java.util.List<NodeGatewayStartedShards>> indexShardsBuilder;
+                    if (indexShards == null) {
+                        indexShardsBuilder = new IntObjectHashMap<>();
+                    } else {
+                        indexShardsBuilder = new IntObjectHashMap<>(indexShards);
+                    }
+                    java.util.List<NodeGatewayStartedShards> shardStoreStatuses = indexShardsBuilder.get(res.shardId.id());
+                    if (shardStoreStatuses == null) {
+                        shardStoreStatuses = new ArrayList<>();
+                    }
+                    for (NodeGatewayStartedShards response : res.responses) {
+                        if (shardExistsInNode(response)) {
+                            shardStoreStatuses.add(response);
+                        }
+                    }
+                    indexShardsBuilder.put(res.shardId.id(), shardStoreStatuses);
+                    shardsResponses.put(res.shardId.getIndex(), indexShardsBuilder);
+                    for (FailedNodeException failure : res.failures) {
+                        failures.add(new IndicesShardsStoresResponse.Failure(failure.nodeId(), res.shardId.getIndex(), res.shardId.id(), failure.getCause()));
+                    }
+                }
+                ImmutableOpenMap.Builder<String, ImmutableOpenIntMap<java.util.List<IndicesShardsStoresResponse.ShardStoreStatus>>> indicesShardStoreStatuses = ImmutableOpenMap.builder();
+                for (Map.Entry<String, IntObjectHashMap<java.util.List<NodeGatewayStartedShards>>> entry : shardsResponses.entrySet()) {
+                    String index = entry.getKey();
+                    ImmutableOpenIntMap.Builder<java.util.List<IndicesShardsStoresResponse.ShardStoreStatus>> shardStoreResponses = ImmutableOpenIntMap.builder();
+                    for (IntObjectCursor<java.util.List<NodeGatewayStartedShards>> shardResponse : entry.getValue()) {
+                        java.util.List<IndicesShardsStoresResponse.ShardStoreStatus> shardStoreStatusList = new ArrayList<>(shardResponse.value.size());
+                        for (int i = 0; i < shardResponse.value.size(); i++) {
+                            NodeGatewayStartedShards res = shardResponse.value.get(i);
+                            final IndicesShardsStoresResponse.ShardStoreStatus.Allocation allocation;
+                            if (res.storeException() == null) {
+                                allocation = isPrimary(index, shardResponse.key, res.getNode())
```

Comment:
FYI - not all shard stores are used .. you can't assume that if there was no error and that it's not a primary it must be a replica.